ext/simplexml/simplexml.c

```diff
@@ -2378,11 +2378,26 @@ PHP_METHOD(SimpleXMLElement, __construct)
 	PHP_LIBXML_RESTORE_GLOBALS(read_file_or_memory);
 
 	if (!docp) {
-		((php_libxml_node_object *)sxe)->document = NULL;
 		zend_throw_exception(zend_ce_exception, "String could not be parsed as XML", 0);
 		RETURN_THROWS();
 	}
 
+	if (!Z_ISUNDEF(sxe->iter.data)) {
```

Comment:
Can you factor this chunk out to a separate method, such that this code is shared with the freeing code for this class?

Reply:
sure